040-valid-identifiers/proposal.md

```diff
@@ -0,0 +1,84 @@
+# Summary
+
+Proposes a fairly limited set of valid characters for use in Concourse identifiers:
+
+* Team names
+* Pipeline names
+* Job names
+* Step names
+* [Var source][var-sources-rfc] names
+
+# Motivation
+
+Concourse does not currently perform much validation for identifiers that can
+be set through the API or in pipeline configs. As a result, users are able to
+name pipelines and jobs with all sorts of special characters.
+
+Allowing arbitrary symbols makes it difficult for Concourse to support semantic
+notation in the CLI and elsewhere in the Concourse UX. For example, the `fly`
+CLI uses `PIPELINE/JOB` syntax, but this becomes ambiguous if your pipeline or
+job names have a `/` present in their name. For example, `foo/bar/baz` could
+either be (`foo/bar`, `baz`) or (`foo`, `bar/baz`).
+
+Allowing whitespace, capitalization, and mixed use of `_` and `-` also results
+in inconsistent naming conventions between Concourse users. A 'deploy to prod'
+job may be called any of the following:
+
+* `deploy to prod`
+* `deploy-to-prod`
+* `deploy_to_prod`
+* `Deploy to prod`
+* `Deploy to Prod`
+
+This variance is largely cosmetic and only makes it difficult for a new
+Concourse user on an existing team to predict the name of a given job. It also
+results in an inconsistent UX when using Concourse.
+
+# Proposal
+
+This proposal is to dramatically reduce the allowed character set for Concourse
+identifiers, in the spirit of simplicity and consistency across Concourse
+usage.
+
+The following characters are to be allowed:
+
+* Lowercase Unicode letters.
```

Comment:
Hmm, my intent in allowing any lowercase Unicode letter over just `a-z` is to support our multilingual friends, however I just realized that it's hard to enforce "lowercase" when some languages don't have casing at all.

e.g.:

```ruby
irb(main):001:0> /\p{Ll}/ =~ "ひらがな"
=> nil
irb(main):002:0> /\p{L}/ =~ "ひらがな"
=> 0
irb(main):003:0>
```

Of course, we could match all these specific scripts by identifying their Unicode block:

```ruby
irb(main):003:0> /\p{Hiragana}/ =~ "ひらがな"
=> 0
```

...but that seems pretty high-maintenance.

I see two paths forward here:

* Enforce `a-z`. Do people feel strongly about this? I don't think I've ever seen an example of e.g. a Japanese job name.
* Allow uppercase letters and permit this level of inconsistency.
  * If we explicitly allow this we should be sure to not allow e.g. `foO` and `Foo` to both exist. I'm not sure what the situation is now.

Reply:
Resolving this by allowing all cases and just converting to lowercase.